Web processes should only commit history state for navigating frames and their children

charliewolfe opened the pull request:
#### 9b13b53b41a196d0fa64076a7d0239843b4eec09
<pre>
Web processes should only commit history state for navigating frames and their children
<a href="https://bugs.webkit.org/show_bug.cgi?id=285224">https://bugs.webkit.org/show_bug.cgi?id=285224</a>
<a href="https://rdar.apple.com/142134339">rdar://142134339</a>

Reviewed by Alex Christensen.

Currently, when a frame navigates, the web process creates history state for the entire frame tree to be
committed to the UI process. This needs to change with site isolation because a navigating iframe may
have parent frames that are out-of-process, which we cannot synchronously create history state for.
Instead, when a frame navigates, the web process should only create and commit history state for the
navigating frame and its children. If the navigating frame is a child frame, we can use the current item
in the UI process&apos;s back/forward list to populate the history state for its ancestors.

On navigation, only frames that navigate will be assigned a new WebBackForwardListFrameItem, while the
history item tree created for the new navigation will always receive a new WebBackForwardListItem. This
allows the UI process to correctly update the history for isolated iframes that did not navigate but had
a sibling in another process that did.

No change in behavior with site isolation disabled (hopefully). More details below.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::BackForwardCache::addIfCacheable):
(WebCore::BackForwardCache::take):
(WebCore::BackForwardCache::get):
(WebCore::BackForwardCache::remove):
* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::addItem):
(WebCore::BackForwardController::setChildItem):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/BackForwardFrameItemIdentifier.h: Added.
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::HistoryItem):
(WebCore::HistoryItem::isInBackForwardCache const):
(WebCore::HistoryItem::hasCachedPageExpired const):
(WebCore::HistoryItem::shouldDoSameDocumentNavigationTo const):
(WebCore::HistoryItem::logString const):
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::create):
(WebCore::HistoryItem::itemID const):
(WebCore::HistoryItem::frameItemID const):
(WebCore::HistoryItem::operator== const):
(WebCore::HistoryItem::identifier const): Deleted.
Assign HistoryItems a BackForwardFrameItemIdentifier, which is unique for each frame in a history item
tree. BackForwardItemIdentifier now only identifies history item trees.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::createHistoryItemTree const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::restoreDocumentState):
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::goToItemForNavigationAPI):
(WebCore::HistoryController::updateForRedirectWithLockedBackForwardList):
(WebCore::HistoryController::createItem):
(WebCore::HistoryController::createItemTree):
(WebCore::HistoryController::itemsAreClones):
(WebCore::HistoryController::updateBackForwardListClippedAtTarget):
Call createItemTree on the navigating frame instead of the root frame. Disabled for WebKit1.

(WebCore::HistoryController::pushState):
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/loader/NavigationAction.cpp:
(WebCore::NavigationAction::setTargetBackForwardItem):
(WebCore::NavigationAction::setSourceBackForwardItem):
* Source/WebCore/loader/NavigationScheduler.cpp:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearPreviousItemFromAllPages):
(WebCore::Page::goToItem):
(WebCore::Page::goToItemForNavigationAPI):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/ProcessQualified.serialization.in:
* Source/WebKit/Shared/SessionState.cpp:
(WebKit::FrameState::FrameState):
(WebKit::FrameState::copy):
(WebKit::FrameState::replaceChildFrameState):
* Source/WebKit/Shared/SessionState.h:
* Source/WebKit/Shared/SessionState.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebBackForwardListFrameItem.cpp:
(WebKit::WebBackForwardListFrameItem::create):
(WebKit::WebBackForwardListFrameItem::WebBackForwardListFrameItem):
(WebKit::WebBackForwardListFrameItem::~WebBackForwardListFrameItem):
(WebKit::WebBackForwardListFrameItem::allItems):
(WebKit::WebBackForwardListFrameItem::itemForID):
(WebKit::WebBackForwardListFrameItem::url const):
(WebKit::WebBackForwardListFrameItem::childItemForTarget):
(WebKit::WebBackForwardListFrameItem::setChild):
(WebKit::WebBackForwardListFrameItem::mainFrame):
(WebKit::WebBackForwardListFrameItem::protectedMainFrame):
(WebKit::WebBackForwardListFrameItem::setFrameState):
(WebKit::WebBackForwardListFrameItem::identifier const): Deleted.
* Source/WebKit/Shared/WebBackForwardListFrameItem.h:
(WebKit::WebBackForwardListFrameItem::identifier const):
* Source/WebKit/Shared/WebBackForwardListItem.cpp:
(WebKit::WebBackForwardListItem::create):
(WebKit::WebBackForwardListItem::WebBackForwardListItem):
(WebKit::WebBackForwardListItem::~WebBackForwardListItem):
(WebKit::WebBackForwardListItem::itemIsInSameDocument const):
(WebKit::WebBackForwardListItem::itemIsClone):
(WebKit::WebBackForwardListItem::navigatedFrameState const):
(WebKit::WebBackForwardListItem::mainFrameState const):
(WebKit::WebBackForwardListItem::originalURL const):
(WebKit::WebBackForwardListItem::url const):
(WebKit::WebBackForwardListItem::title const):
(WebKit::WebBackForwardListItem::wasCreatedByJSWithoutUserInteraction const):
(WebKit::WebBackForwardListItem::setWasRestoredFromSession):
(WebKit::WebBackForwardListItem::navigatedFrameItem const):
(WebKit::WebBackForwardListItem::protectedNavigatedFrameItem const):
(WebKit::WebBackForwardListItem::mainFrameItem const):
(WebKit::WebBackForwardListItem::loggingString):
(WebKit::WebBackForwardListItem::itemID const): Deleted.
(WebKit::WebBackForwardListItem::setRootFrameState): Deleted.
(WebKit::WebBackForwardListItem::rootFrameState const): Deleted.
(WebKit::WebBackForwardListItem::protectedRootFrameItem): Deleted.
(WebKit::WebBackForwardListItem::setParentFromItem): Deleted.
This function is no longer needed because items for remote child frame navigations now have their parents
set during WebBackForwardListItem construction.

* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::identifier const):
(WebKit::WebBackForwardListItem::setNavigatedFrameID): Deleted.
(WebKit::WebBackForwardListItem::rootFrameItem): Deleted.
Replace rootFrameState() and rootFrameItem() with mainFrameState() and mainFrameItem(), as each item
should now have a WebBackForwardListFrameItem for the main frame.

* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::Navigation):
(API::Navigation::targetItem const):
(API::Navigation::loggingString const):
Changed m_targetItem to hold a WebBackForwardListFrameItem, so we can use it to navigate the correct
frame in ProvisionalPageProxy::goToBackForwardItem().

* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::create):
(API::Navigation::protectedTargetItem const):
(API::Navigation::targetFrameItem const):
(API::Navigation::targetItem const): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.mm:
(-[WKBackForwardListItem _scrollPosition]):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::goToBackForwardItem):
(WebKit::ProvisionalPageProxy::backForwardAddItem):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardCache.cpp:
(WebKit::WebBackForwardCache::addEntry):
(WebKit::WebBackForwardCache::removeEntry):
(WebKit::WebBackForwardCache::takeSuspendedPage):
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::itemForID):
(WebKit::WebBackForwardList::addItem):
(WebKit::WebBackForwardList::goToItemInternal):
(WebKit::WebBackForwardList::backForwardListState const):
(WebKit::setBackForwardItemIdentifiers):
(WebKit::WebBackForwardList::restoreFromState):
completeFrameStateForNavigation() needs to refer to the current back/forward item to correctly complete
frame state for navigating child frame items. So, m_provisionalIndex is cleared earlier, and
m_currentIndex now points to the last entry in m_entries during restoration of the back/forward list.

(WebKit::WebBackForwardList::didRemoveItem):
(WebKit::setBackForwardItemIdentifier):
(WebKit::WebBackForwardList::completeFrameStateForNavigation):
This function completes the FrameState for navigating child frames. The BackForwardItemIdentifiers from
the current item are regenerated when set on the new item, as the web process must be able to uniquely
address each WebBackForwardListFrameItem.

* Source/WebKit/UIProcess/WebBackForwardList.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::loadData):
This is needed to pass the message checks in WebPageProxy::backForwardAddItemShared, since the added item
can now be a child frame URL.

* Source/WebKit/UIProcess/WebNavigationState.cpp:
(WebKit::WebNavigationState::createBackForwardNavigation):
* Source/WebKit/UIProcess/WebNavigationState.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::suspendCurrentPageIfPossible):
(WebKit::WebPageProxy::launchProcessForReload):
(WebKit::WebPageProxy::goForward):
(WebKit::WebPageProxy::goBack):
(WebKit::WebPageProxy::goToBackForwardItem):
If a frame no longer exists, such as when the web process crashes and the page is reloaded, we cannot
instruct the web process to load a history item into it and should instead send the entire frame tree.
Not doing this would mean loading an iframe URL in the main frame.

(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::backForwardAddItem):
(WebKit::WebPageProxy::backForwardAddItemShared):
(WebKit::WebPageProxy::backForwardSetChildItem):
(WebKit::WebPageProxy::backForwardUpdateItem):
(WebKit::WebPageProxy::backForwardClearProvisionalItem):
(WebKit::WebPageProxy::backForwardItemAtIndex):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldSendPendingMessage):
* Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
(WebKit::ViewGestureController::beginSwipeGesture):
* Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.cpp:
(WebKit::toFrameState):
(WebKit::applyFrameState):
(WebKit::toHistoryItem):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::shouldGoToHistoryItem const):
(WebKit::WebLocalFrameLoaderClient::createHistoryItemTree const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::addItem):
The parameter indicating which frame is being navigated can be removed, since the frame at the top of the
tree being committed will now always be the navigating frame.

(WebKit::WebBackForwardListProxy::setChildItem):
(WebKit::WebBackForwardListProxy::goToItem):
(WebKit::WebBackForwardListProxy::goToProvisionalItem):
(WebKit::WebBackForwardListProxy::clearProvisionalItem):
(WebKit::WebBackForwardListProxy::containsItem const):
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::goToBackForwardItem):
* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::addItem):
* Source/WebKitLegacy/mac/History/WebBackForwardList.mm:
(-[WebBackForwardList addItem:]):
(-[WebBackForwardList setToMatchDictionaryRepresentation:]):
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::createHistoryItemTree const):
Creating history item trees from the navigating frame doesn&apos;t make sense in WebKit1, so disable it.

* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _loadBackForwardListFromOtherView:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, GoBackToNestedIframeCreatedAfterNavigatingSibling)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView secondChildFrame]):

Canonical link: <a href="https://commits.webkit.org/288409@main">https://commits.webkit.org/288409@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9dc8a16c48ff760d73867190193de6fd278ebcda

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83079 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2720 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37372 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88188 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34125 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85171 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2795 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10642 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64685 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22441 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86134 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2045 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75534 "Found 2 new API test failures: TestWebKitAPI.WKWebExtensionAPITabs.ConnectToSpecificFrame, TestWebKitAPI.WKWebExtensionAPITabs.SendMessageFromBackgroundToSpecificFrame (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44966 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/82511 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1953 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29738 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33159 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73122 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30433 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89554 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10361 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7484 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73118 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10591 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71353 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72344 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16523 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15255 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12850 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10314 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10186 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13651 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11955 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2526,28 +2532,29 @@ RefPtr<API::Navigation> WebPageProxy::goToBackForwardItem(WebBackForwardListItem
 
-    auto publicSuffix = WebCore::PublicSuffixStore::singleton().publicSuffix(URL(item.url()));
-    process->send(Messages::WebPage::GoToBackForwardItem({ navigation->navigationID(), frameItem.rootFrame().copyFrameStateWithChildren(), frameLoadType, ShouldTreatAsContinuingLoad::No, std::nullopt, m_lastNavigationWasAppInitiated, std::nullopt, WTFMove(publicSuffix), { } }), webPageIDInProcess(process));
+    Ref frameState = WebFrameProxy::webFrame(frameItem.frameID()) ? frameItem.copyFrameStateWithChildren() : item->mainFrameState();
```
I don't think I understand this condition or when it would fail.

charliewolfe replied:
If a frame no longer exists for some reason, such as when the web process crashes and the page is reloaded, we cannot instruct the web process to load a history item into it and should instead send the entire frame tree. I’ll mention this in the commit message.

achristensen07 commented on Source/WebKit/UIProcess/WebBackForwardList.cpp:
```diff
@@ -68,11 +68,7 @@ WebBackForwardListItem* WebBackForwardList::itemForID(const BackForwardItemIdent
-        return nullptr;
-
-    auto* item = frameItem->backForwardListItem();
+    auto* item = WebBackForwardListItem::itemForID(identifier);
```
RefPtr

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -325,7 +326,8 @@ void ProvisionalPageProxy::goToBackForwardItem(API::Navigation& navigation, WebB
     URL itemURL { item.url() };
-    page->maybeInitializeSandboxExtensionHandle(process(), itemURL, item.resourceDirectoryURL(), true, [weakThis = WeakPtr { *this }, itemURL, frameState = item.rootFrameState(), navigationLoadType = *navigation.backForwardFrameLoadType(), shouldTreatAsContinuingLoad, websitePoliciesData = WTFMove(websitePoliciesData), existingNetworkResourceLoadIdentifierToResume = WTFMove(existingNetworkResourc ...
+    Ref frameState = navigation.targetFrameItem() ? navigation.targetFrameItem()->copyFrameStateWithChildren() : item.mainFrameState();
+    page->maybeInitializeSandboxExtensionHandle(process(), itemURL, item.resourceDirectoryURL(), true, [weakThis = WeakPtr { *this }, itemURL, frameState = WTFMove(frameState), navigationLoadType = *navigation.backForwardFrameLoadType(), shouldTreatAsContinuingLoad, websitePoliciesData = WTFMove(websitePoliciesData), existingNetworkResourceLoadIdentifierToResume = WTFMove(existingNetworkResourceL ...
```
This capture list may be worth putting each item on its own line.

achristensen07 commented on Source/WebKit/UIProcess/API/APINavigation.h:
```diff
@@ -30,6 +30,7 @@
 #include "FrameInfoData.h"
 #include "NavigationActionData.h"
 #include "ProcessThrottler.h"
+#include "WebBackForwardListFrameItem.h"
```
Can a forward declaration suffice?

charliewolfe replied:
Yes

achristensen07 commented on Source/WebKit/Shared/WebBackForwardListFrameItem.cpp:
```diff
@@ -32,47 +32,53 @@
 }
 
-BackForwardItemIdentifier WebBackForwardListFrameItem::identifier() const
+const String& WebBackForwardListFrameItem::url() const
```
urlString?

charliewolfe replied:
This matches `WebBackForwardListItem::url()`. I can change both but maybe not in this PR?

achristensen07 replied:
no, this is fine.

achristensen07 commented on Source/WebKit/Shared/WebBackForwardListFrameItem.cpp:
```diff
@@ -32,47 +32,53 @@
+
+BackForwardFrameItemIdentifier WebBackForwardListFrameItem::identifier() const
+{
+    return *m_frameState->frameItemID;
```
Can we express in the type that this is always present without dereferencing an optional?

achristensen07 commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -7771,6 +7771,20 @@ UnifyDamagedRegions:
      WebKit:
        default: false
 
+UpdateHistoryByNavigatingFrame:
```
Could we use a FrameLoaderClient call instead of adding a preference?  This is just to distinguish WebKitLegacy from modern WebKit, right?  We don't actually need this to be runtime configurable.

charliewolfe committed:
Web processes should only commit history state for navigating frames and their children
https://bugs.webkit.org/show_bug.cgi?id=285224
rdar://142134339

Reviewed by Alex Christensen.

Currently, when a frame navigates, the web process creates history state for the entire frame tree to be
committed to the UI process. This needs to change with site isolation because a navigating iframe may
have parent frames that are out-of-process, which we cannot synchronously create history state for.
Instead, when a frame navigates, the web process should only create and commit history state for the
navigating frame and its children. If the navigating frame is a child frame, we can use the current item
in the UI process's back/forward list to populate the history state for its ancestors.

On navigation, only frames that navigate will be assigned a new WebBackForwardListFrameItem, while the
history item tree created for the new navigation will always receive a new WebBackForwardListItem. This
allows the UI process to correctly update the history for isolated iframes that did not navigate but had
a sibling in another process that did.

No change in behavior with site isolation disabled (hopefully). More details below.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::BackForwardCache::addIfCacheable):
(WebCore::BackForwardCache::take):
(WebCore::BackForwardCache::get):
(WebCore::BackForwardCache::remove):
* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::addItem):
(WebCore::BackForwardController::setChildItem):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/BackForwardFrameItemIdentifier.h: Added.
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::HistoryItem):
(WebCore::HistoryItem::isInBackForwardCache const):
(WebCore::HistoryItem::hasCachedPageExpired const):
(WebCore::HistoryItem::shouldDoSameDocumentNavigationTo const):
(WebCore::HistoryItem::logString const):
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::create):
(WebCore::HistoryItem::itemID const):
(WebCore::HistoryItem::frameItemID const):
(WebCore::HistoryItem::operator== const):
(WebCore::HistoryItem::identifier const): Deleted.
Assign HistoryItems a BackForwardFrameItemIdentifier, which is unique for each frame in a history item
tree. BackForwardItemIdentifier now only identifies history item trees.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::createHistoryItemTree const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::restoreDocumentState):
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::goToItemForNavigationAPI):
(WebCore::HistoryController::updateForRedirectWithLockedBackForwardList):
(WebCore::HistoryController::createItem):
(WebCore::HistoryController::createItemTree):
(WebCore::HistoryController::itemsAreClones):
(WebCore::HistoryController::updateBackForwardListClippedAtTarget):
Call createItemTree on the navigating frame instead of the root frame. Disabled for WebKit1.

(WebCore::HistoryController::pushState):
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/loader/NavigationAction.cpp:
(WebCore::NavigationAction::setTargetBackForwardItem):
(WebCore::NavigationAction::setSourceBackForwardItem):
* Source/WebCore/loader/NavigationScheduler.cpp:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearPreviousItemFromAllPages):
(WebCore::Page::goToItem):
(WebCore::Page::goToItemForNavigationAPI):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/ProcessQualified.serialization.in:
* Source/WebKit/Shared/SessionState.cpp:
(WebKit::FrameState::FrameState):
(WebKit::FrameState::copy):
(WebKit::FrameState::replaceChildFrameState):
* Source/WebKit/Shared/SessionState.h:
* Source/WebKit/Shared/SessionState.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebBackForwardListFrameItem.cpp:
(WebKit::WebBackForwardListFrameItem::create):
(WebKit::WebBackForwardListFrameItem::WebBackForwardListFrameItem):
(WebKit::WebBackForwardListFrameItem::~WebBackForwardListFrameItem):
(WebKit::WebBackForwardListFrameItem::allItems):
(WebKit::WebBackForwardListFrameItem::itemForID):
(WebKit::WebBackForwardListFrameItem::url const):
(WebKit::WebBackForwardListFrameItem::childItemForTarget):
(WebKit::WebBackForwardListFrameItem::setChild):
(WebKit::WebBackForwardListFrameItem::mainFrame):
(WebKit::WebBackForwardListFrameItem::protectedMainFrame):
(WebKit::WebBackForwardListFrameItem::setFrameState):
(WebKit::WebBackForwardListFrameItem::identifier const): Deleted.
* Source/WebKit/Shared/WebBackForwardListFrameItem.h:
(WebKit::WebBackForwardListFrameItem::identifier const):
* Source/WebKit/Shared/WebBackForwardListItem.cpp:
(WebKit::WebBackForwardListItem::create):
(WebKit::WebBackForwardListItem::WebBackForwardListItem):
(WebKit::WebBackForwardListItem::~WebBackForwardListItem):
(WebKit::WebBackForwardListItem::itemIsInSameDocument const):
(WebKit::WebBackForwardListItem::itemIsClone):
(WebKit::WebBackForwardListItem::navigatedFrameState const):
(WebKit::WebBackForwardListItem::mainFrameState const):
(WebKit::WebBackForwardListItem::originalURL const):
(WebKit::WebBackForwardListItem::url const):
(WebKit::WebBackForwardListItem::title const):
(WebKit::WebBackForwardListItem::wasCreatedByJSWithoutUserInteraction const):
(WebKit::WebBackForwardListItem::setWasRestoredFromSession):
(WebKit::WebBackForwardListItem::navigatedFrameItem const):
(WebKit::WebBackForwardListItem::protectedNavigatedFrameItem const):
(WebKit::WebBackForwardListItem::mainFrameItem const):
(WebKit::WebBackForwardListItem::loggingString):
(WebKit::WebBackForwardListItem::itemID const): Deleted.
(WebKit::WebBackForwardListItem::setRootFrameState): Deleted.
(WebKit::WebBackForwardListItem::rootFrameState const): Deleted.
(WebKit::WebBackForwardListItem::protectedRootFrameItem): Deleted.
(WebKit::WebBackForwardListItem::setParentFromItem): Deleted.
This function is no longer needed because items for remote child frame navigations now have their parents
set during WebBackForwardListItem construction.

* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::identifier const):
(WebKit::WebBackForwardListItem::setNavigatedFrameID): Deleted.
(WebKit::WebBackForwardListItem::rootFrameItem): Deleted.
Replace rootFrameState() and rootFrameItem() with mainFrameState() and mainFrameItem(), as each item
should now have a WebBackForwardListFrameItem for the main frame.

* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::Navigation):
(API::Navigation::targetItem const):
(API::Navigation::loggingString const):
Changed m_targetItem to hold a WebBackForwardListFrameItem, so we can use it to navigate the correct
frame in ProvisionalPageProxy::goToBackForwardItem().

* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::create):
(API::Navigation::protectedTargetItem const):
(API::Navigation::targetFrameItem const):
(API::Navigation::targetItem const): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.mm:
(-[WKBackForwardListItem _scrollPosition]):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::goToBackForwardItem):
(WebKit::ProvisionalPageProxy::backForwardAddItem):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardCache.cpp:
(WebKit::WebBackForwardCache::addEntry):
(WebKit::WebBackForwardCache::removeEntry):
(WebKit::WebBackForwardCache::takeSuspendedPage):
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::itemForID):
(WebKit::WebBackForwardList::addItem):
(WebKit::WebBackForwardList::goToItemInternal):
(WebKit::WebBackForwardList::backForwardListState const):
(WebKit::setBackForwardItemIdentifiers):
(WebKit::WebBackForwardList::restoreFromState):
completeFrameStateForNavigation() needs to refer to the current back/forward item to correctly complete
frame state for navigating child frame items. So, m_provisionalIndex is cleared earlier, and
m_currentIndex now points to the last entry in m_entries during restoration of the back/forward list.

(WebKit::WebBackForwardList::didRemoveItem):
(WebKit::setBackForwardItemIdentifier):
(WebKit::WebBackForwardList::completeFrameStateForNavigation):
This function completes the FrameState for navigating child frames. The BackForwardItemIdentifiers from
the current item are regenerated when set on the new item, as the web process must be able to uniquely
address each WebBackForwardListFrameItem.

* Source/WebKit/UIProcess/WebBackForwardList.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::loadData):
This is needed to pass the message checks in WebPageProxy::backForwardAddItemShared, since the added item
can now be a child frame URL.

* Source/WebKit/UIProcess/WebNavigationState.cpp:
(WebKit::WebNavigationState::createBackForwardNavigation):
* Source/WebKit/UIProcess/WebNavigationState.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::suspendCurrentPageIfPossible):
(WebKit::WebPageProxy::launchProcessForReload):
(WebKit::WebPageProxy::goForward):
(WebKit::WebPageProxy::goBack):
(WebKit::WebPageProxy::goToBackForwardItem):
If a frame no longer exists, such as when the web process crashes and the page is reloaded, we cannot
instruct the web process to load a history item into it and should instead send the entire frame tree.
Not doing this would mean loading an iframe URL in the main frame.

(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::backForwardAddItem):
(WebKit::WebPageProxy::backForwardAddItemShared):
(WebKit::WebPageProxy::backForwardSetChildItem):
(WebKit::WebPageProxy::backForwardUpdateItem):
(WebKit::WebPageProxy::backForwardClearProvisionalItem):
(WebKit::WebPageProxy::backForwardItemAtIndex):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldSendPendingMessage):
* Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
(WebKit::ViewGestureController::beginSwipeGesture):
* Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.cpp:
(WebKit::toFrameState):
(WebKit::applyFrameState):
(WebKit::toHistoryItem):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::shouldGoToHistoryItem const):
(WebKit::WebLocalFrameLoaderClient::createHistoryItemTree const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::addItem):
The parameter indicating which frame is being navigated can be removed, since the frame at the top of the
tree being committed will now always be the navigating frame.

(WebKit::WebBackForwardListProxy::setChildItem):
(WebKit::WebBackForwardListProxy::goToItem):
(WebKit::WebBackForwardListProxy::goToProvisionalItem):
(WebKit::WebBackForwardListProxy::clearProvisionalItem):
(WebKit::WebBackForwardListProxy::containsItem const):
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::goToBackForwardItem):
* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::addItem):
* Source/WebKitLegacy/mac/History/WebBackForwardList.mm:
(-[WebBackForwardList addItem:]):
(-[WebBackForwardList setToMatchDictionaryRepresentation:]):
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::createHistoryItemTree const):
Creating history item trees from the navigating frame doesn't make sense in WebKit1, so disable it.

* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _loadBackForwardListFromOtherView:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, GoBackToNestedIframeCreatedAfterNavigatingSibling)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView secondChildFrame]):

Canonical link: https://commits.webkit.org/288409@main
